### ISSUE TITLE
--exclude versus --pattern

### ISSUE THREAD

**Issue by maethor** [borgbackup/borg#6134]
## Have you checked borgbackup docs, FAQ, and open Github issues?

Yes

## Is this a BUG / ISSUE report or a QUESTION?

QUESTION

I have two questions about --exclude and --pattern.

- First, `--pattern` is marked "experimental" but I cannot find what it really means. Can we use it for production or is it really expected to change?

- Second, I understand very well the difference between patterns `- /proc` and `! /proc`, but I cannot find what `--exclude /proc` is really doing. Is it equivalent to `- /proc`, or `! /proc`? Does it scan the repository, or not?

Thank you. I'm truly sorry if the answers was in the docs or here but I could not find them.

**Comment by ThomasWaldmann** [comment]
1. considering patterns are there since quite a while now, guess we should remove the "experimental" for next release.

2. IIRC `--exclude` is like `"! pattern"` (== has **norecurse** behaviour) - especially since there is no corresponding `--include` (which would be the only reason to recurse into, looking for more stuff to back up)

**Comment by goebbe** [comment]
Just a related comment. I am not sure if this case is even officially supported. At least I didn't find this in the documentation. 
 
It is also possible to combine `--pattern` and `--exclude`. 

I used the following command with `--exclude` to exclude all hidden files and directories from my backup. 
It was sufficient to add a `--pattern=+` line **before** the `--exlude` line (order matters) to add the backup of my .thunderbird file. 

```
borg create                                     \
    --pattern=+/home/username/.thunderbird       \
    --exclude '*/[.]*'                          \
    ::'{hostname}-{now}'                        \
    /home/username                               \
```

This might be obvious to the developers. But it was certainly not obvious to me.  :-)
If this is official supported, I recommend adding an example to the documentation.

**Comment by ThomasWaldmann** [comment]
Well, I guess I'ld rather stay with using one system than mixing two, especially since patterns are more powerful than the `--exclude`.

**Comment by goebbe** [comment]
Fair enough. :-)   
My problem was, that I had quite a list of `--exclude` commands that "did just work" and I didn't have time to switch to pattern. 

What I really like about `--exclude`  is that it is so expressive about its meaning.  By the way, I discovered `--pattern` when looking for a (non existing)  `--include` equivalent to  `--exclude`.     :-)

**Comment by goebbe** [comment]
Perhaps the following case is interesting for the FAQ or examples or documentation: 

When I tried to switch from `--exclude` to `--pattern`:
The following exclude command was in my old borgbackup.sh-script: 
`  --exclude '*.iso'     `          this is sufficient to exclude all .iso files from being backed up. 

I ran into several stumbling blocks, when I tried to replicate the same behaviour, using `--pattern`:  
`  --pattern=-'*.iso'     `      this did not work. My .iso files (from subfolders) have been transferred to the backup server. 
`  --pattern=- '*.iso'    `     this did not work, there was an error message caused by the space after the `=-`.

However, each of the following worked: 
`  --pattern=-'fm:*.iso'       `   this did work. Just put the `fm:` before the old pattern - `fm:` is the default pattern in `--exclude`.
or: 
`  --pattern=-'**/*.iso'        `    this did work. This uses the `sh:` pattern. Which is default in `--pattern`.

**Comment by ThomasWaldmann** [comment]
todo: check if we need to add some more examples / clarifications.

**Comment by goebbe** [comment]
Perhaps it would be a good idea to integrate the borg-patterns **man page** into the documentation? And add additional examples? 

References: 
Man page: https://manpages.debian.org/testing/borgbackup/borg-patterns.1.en.html
borg documentation: https://borgbackup.readthedocs.io/en/stable/usage/general.html

**Comment by ThomasWaldmann** [comment]
IIRC that is in the docs, did you use the search function?

**Comment by goebbe** [comment]
Sorry for the noise. I looked under `Usage` and `borg create`, but I didn't find it. 
The only reference I found there is to use `borg help patterns` - perhaps there could be a link to the correct help page?  

Searching for "pattern": gives a long list with different topics - it is not obvious from the page-titles which one would provide an overview for patterns. I actually tried some of them, but didn't find the relevant information. 

Actually, it would have been the first hit, which is called "Miscellaneous Help" - the first part of "Miscalleneous Help" is basically about patterns. 
Perhaps one could create an own page for pattern and call it "borg patterns", to make it more visible? 

So the problem was (partly) my wrong expectations - I did expect something more specific to turn up, something like "patterns and exclude".
